### PR TITLE
Stabilize `rustdoc::bare_urls` lint

### LINKED ISSUE #77501: rustdoc: Suggest to turn 'http...' into an automatic link
It would be nice if a raw hyperlink (with no markup) in a doc comment were rendered out as an anchor tag, à la GitHub. Otherwise you have to write
```rust
/// Check out this link: [https://example.com](https://example.com)
fn foo();
```
This would maximize readability in the code while slightly improving the generated HTML.

### LINKED ISSUE #83598: Renaming a tool lint to another tool lint does not apply the warning
Similar to https://github.com/rust-lang/rust/issues/82615, but the source lint is a tool lint this time.

I tried this code:

```rust
#![deny(rustdoc::non_autolinks)]
//! http://example.com
```

and renamed `non_autolinks` in a fork of rust-lang/rust.

I expected to see this happen: Rustdoc warns that the lint is renamed and still applies the lint.

Instead, this happened: Rustdoc warns that the lint is renamed, but does not apply the lint.

```
warning: lint `rustdoc::non_autolinks` has been renamed to `rustdoc::bare_urls`
  --> /home/joshua/rustc/src/test/rustdoc-ui/url-improvements.rs:1:9
   |
LL | #![deny(rustdoc::non_autolinks)]
   |         ^^^^^^^^^^^^^^^^^^^^^^ help: use the new name: `rustdoc::bare_urls`
   |
   = note: `#[warn(renamed_and_removed_lints)]` on by default
warning: this URL is not a hyperlink
  --> /home/joshua/rustc/src/test/rustdoc-ui/url-improvements.rs:2:5
   |
LL | /// https://example.com
   |     ^^^^^^^^^^^^^^^^^^^^^^^ help: use an automatic link instead: `<https://example.com>`
```

### Meta

Version: https://github.com/rust-lang/rust/commit/59d7295c321809740961996b7369f27fa3c3c024

### PR DESCRIPTION
**Pull request by jyn514**
Closes https://github.com/rust-lang/rust/issues/77501. Closes https://github.com/rust-lang/rust/issues/83598.

### CODE REVIEW

**Comment by jyn514** [comment]
@rfcbot fcp merge

This lint has been on nightly for about three months without issues. It seems
useful and in-scope, so this proposes stabilizing it.

## Potential concerns

- `non_autolinks` is kind of an odd name. Maybe we should rename it to
  `missing_links` or something?

**Comment by Manishearth** [comment]
I would call it `bare_link` or something, the standard lint guidance is to say what it's linting against in the name

**Comment by jyn514** [comment]
> I would call it `bare_link` or something, the standard lint guidance is to say what it's linting against in the name

Sorry, I opened the FCP before realizing this does more than warn about bare links - this also warns about `[x](x)`, which isn't a bare link.

**Comment by camelid** [comment]
>     * `non_autolinks` is kind of an odd name. Maybe we should rename it to
>       `missing_links` or something?

`non_linkified_url`? `unlinked_url`?

**Comment by jyn514** [comment]
> > * `non_autolinks` is kind of an odd name. Maybe we should rename it to
> >   `missing_links` or something?
> 
> `non_linkified_url`? `unlinked_url`?

Both of those are less clear than `non_autolinks` that `[x](x)` also warns IMO.

**Comment by jyn514** [comment]
We don't *have* to change the name, btw, I think it's ok as is, just a little odd.

**Comment by camelid** [comment]
IMO `non_autolinks` doesn't really make sense unless you already know what it does though. Marking the concern: @rfcbot concern naming

**Comment by camelid** [comment]
@rfcbot concern naming

(rfcbot seems pickier about when it recognizes a command)

**Comment by ogoffart** [comment]
Maybe the naming problem suggests that this could be split into two lints:
 * `bare_url`
 * `redundant_link`

The rationale to split into different lint is that one is just a style issue whose suggestion can always be applied and makes no difference in the generated docs; while the other is an actual problem making a real difference in the doc, and might have false positive if a clickable link was really not intended.

**Comment by Nemo157** [comment]
Should this be delayed till #80527 merges, so that `non_autolinks` can start as a scoped lint and doesn't need to be registered as removed?

Though, it looks like the `non_autolinks` lint name has already leaked onto stable; `#[warn(non_autolinks)]` doesn't cause a "unknown lint" warning. Searching github I see a handful of users, mostly `allow(non_autolinks)` so any renaming may have to be done carefully if we want to not break their builds (though since it's just a lint change I think we're fine to just change it under the stability policy).

**Comment by Manishearth** [comment]
> Sorry, I opened the FCP before realizing this does more than warn about bare links - this also warns about `[x](x)`, which isn't a bare link.

Honestly I don't think we should care about the latter. But yeah, this points to needing two lints.

**Comment by LeSeulArtichaut** [comment]
> @rfcbot concern unnecessary-half-of-lint

(Looks like @rfcbot missed the command)

**Comment by Manishearth** [comment]
@rfcbot concern unnecessary-half-of-lint

**Comment by jyn514** [comment]
I updated this to remove the "unneceesary long form lint". I'll wait until https://github.com/rust-lang/rust/pull/80527 is merged before messing with renaming.

**Comment by Manishearth** [comment]
The lint name should also be updated.

**Comment by jyn514** [comment]
> Searching github I see a handful of users, mostly allow(non_autolinks) so any renaming may have to be done carefully if we want to not break their builds (though since it's just a lint change I think we're fine to just change it under the stability policy).

Renamed lints are only warnings, warnings are never breaking changes.

> The lint name should also be updated.

Done.  @camelid @Manishearth could you resolve your concerns (or explain why you think they're still relevant)?

**Comment by jyn514** [comment]
Also, I'm happy to break everything but stabilization into a separate PR so it can bake on nightly, if you think that would be helpful. I do want to note this is the start of the release cycle though, so we have 12 weeks to revert if we need to

**Comment by Manishearth** [comment]
@rfcbot resolve naming
@rfcbot resolve unnecessary-half-of-lint

**Comment by Manishearth** [comment]
@rfcbot reviewed

**Comment by jyn514** [comment]
Hmm, rfcbot seems buggy, it didn't pick up on any of those commands :/

**Comment by Manishearth** [comment]
@rfcbot resolve naming
@rfcbot resolve unnecessary-half-of-lint
@rfcbot reviewed

**Comment by Manishearth** [comment]
@rfcbot resolve naming

**Comment by jyn514** [comment]
Nope, still not working :/ cc @rust-lang/infra, rfcbot isn't responding to commands.

**Comment by Manishearth** [comment]
@rfcbot reviewed

**Comment by Manishearth** [comment]
@rfcbot resolve unnecessary-half-of-lint

**Comment by Mark-Simulacrum** [comment]
Investigation seems to point at there not being an active (proposed) FCP in this issue, as it was closed/reopened a bit ago, and that is treated as a cancellation by rfcbot.

**Comment by jyn514** [comment]
> Investigation seems to point at there not being an active (proposed) FCP in this issue, as it was closed/reopened a bit ago, and that is treated as a cancellation by rfcbot.

Is there a way to tell rfcbot it should be treated as active?

**Comment by Manishearth** [comment]
@rfcbot fcp merge

like that^

**Comment by Manishearth** [comment]
The old concerns are resolved anyway, so ... whatever; I ticked your box.

**Comment by Nemo157** [comment]
Should we update the title to the new lint name for searchability?

File: src/librustdoc/lint.rs [hunk 1]
```diff
@@ -148,14 +148,13 @@ declare_rustdoc_lint! {
 }
 
 declare_rustdoc_lint! {
-    /// The `non_autolinks` lint detects when a URL could be written using
-    /// only angle brackets. This is a `rustdoc` only lint, see the
-    /// documentation in the [rustdoc book].
+    /// The `non_autolinks` lint detects when a URL is not a hyperlink.
```
Comment (Nemo157) [inline src/librustdoc/lint.rs:151]:
```suggestion
    /// The `bare_urls` lint detects when a URL is not a hyperlink.
```

Or should it be `rustdoc::bare_urls` in this prose?

Comment (jyn514) [reply src/librustdoc/lint.rs:151]: All the other lints use the bare name; I think I'd rather have it consistent. Happy to change it in a separate PR if you think it's useful.

Also right now the doc-comments aren't used for anything; there's an issue open to stop duplicating the same text in librustdoc and `src/doc/rustdoc`: https://github.com/rust-lang/rust/issues/81305

File: src/doc/rustdoc/src/lints.md [hunk 2]
```diff
@@ -294,40 +294,33 @@ warning: unclosed HTML tag `h1`
 warning: 2 warnings emitted
 ```
 
-## non_autolinks
+## bare_urls
 
-This lint is **nightly-only** and **warns by default**. It detects links which
-could use the "automatic" link syntax. For example:
+This lint is **nightly-only** and **warns by default**. It detects URLs which are not links.
```
Comment (camelid) [inline src/doc/rustdoc/src/lints.md:299]: nit: After this PR, it won't be nightly-only.

```suggestion
This lint is **warn-by-default**. It detects URLs which are not links.
```

**Review by jyn514 — COMMENTED** [review]
~~I really hate how github makes the 'commit batch suggestions' and 'add a review button' look the same~~

**Comment by GuillaumeGomez** [comment]
Is there anything else to be done here? From the conversation, it seems to be ready to go.

**Comment by jyn514** [comment]
@GuillaumeGomez it's still in FCP: https://github.com/rust-lang/rust/pull/81764#issuecomment-809563953

**Pull request merged**